lib/tree.rb

```diff
@@ -16,51 +16,225 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(n)
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # def add(key, value)
+  #   new_node = TreeNode.new(key, value)
+  #   current = @root
+  #   if current.nil?
+  #     current = new_node
+  #   else
+  #     done = false
+  #     until done
+  #       if key <= current.key
+  #         if !current.left.nil?
+  #           current = current.left
+  #         else
+  #           current.left = new_node
+  #           done = true
+  #         end
+  #       else
+  #         if !current.right.nil?
+  #           current = current.right
+  #         else
+  #           current.right = new_node
+  #           done = true
+  #         end
+  #       end
+  #     end
+  #   end
+  #   return current
+  # end
+
+  # Time Complexity: O(log n)
+  # Space Complexity: O(n)
   def find(key)
```

Comment:
Since the number of recursive calls never go above the depth of the tree, it's O(log n) for space complexity due to the recursive call stack.

This of course assumes the tree is balanced.